core/src/main/scala/kafka/server/AutoTopicCreationManager.scala

```diff
@@ -91,11 +91,14 @@ class DefaultAutoTopicCreationManager(
   config: KafkaConfig,
   metadataCache: MetadataCache,
   channelManager: Option[BrokerToControllerChannelManager],
-  adminManager: ZkAdminManager,
-  controller: KafkaController,
+  adminManager: Option[ZkAdminManager],
+  controller: Option[KafkaController],
   groupCoordinator: GroupCoordinator,
   txnCoordinator: TransactionCoordinator
 ) extends AutoTopicCreationManager with Logging {
+  if (controller.isEmpty && channelManager.isEmpty) {
+    throw new IllegalArgumentException("Must supply a channel manager if not supplying a controller")
+  }
```

Comment:
Are these two arguments mutually exclusive? If so, maybe we can use `Either` instead of two options.

Reply:
Yeah, actually a lot more checking could be done here since it's also `adminManager` that is optional.  We created `ZkSupport` and `RaftSupport` for `KafkaApis` and I thought that perhaps the same thing could be done here, but it felt like overkill.  These are not mutually exclusive, actually -- all three could be supplied when using ZooKeeper.

Reply:
I agree it would be nice to do something like that.  This is probably OK for now, though...